url5e6a51063b_nuigroup_hci_for_java_tgz-pJ8-nui_squirt_demo_DemoSquirtApplicationJ8/src/nui/squirt/component/TextField.java

```diff
@@ -0,0 +1,347 @@
+		return caretIndex;
+	}
+
+	public void setCaretIndex(int caretIndex) {
+		if (caretIndex < 0) caretIndex = 0;
+		if (caretIndex > getText().length()) caretIndex = getText().length();
+		this.caretIndex = caretIndex;
+	}
+
+	public int getSecondaryCaretIndex() {
+		return secondaryCaretIndex;
+	}
+
+	public void setSecondaryCaretIndex(int secondaryCaretIndex) {
+		if (secondaryCaretIndex < 0) secondaryCaretIndex = 0;
+		if (secondaryCaretIndex > getText().length()) secondaryCaretIndex = getText().length();
+		this.secondaryCaretIndex = secondaryCaretIndex;
+	}
+
+	@Override
+	public Color getFillColor() {
+		return TEXT_FIELD_FILL_COLOR;
+	}
+	
+	@Override
+	public Color getStrokeColor() {
+		return TEXT_FIELD_BORDER_COLOR;
+	}
+	
+	@Override
+	public float getStrokeWeight() {
+		return TEXT_FIELD_BORDER_WIDTH;
+	}
+	
+	@Override
+	public void preRender(PApplet p) {
+		super.preRender(p);
+		setHeight(p.textAscent()+p.textDescent()+2*TEXT_FIELD_TEXT_PADDING);
+	}
+	
+	@Override
+	public void render(PApplet p) {
+		super.render(p);
+		
+		if (textCanvas == null) {
+			textCanvas = p.createGraphics((int) getWidth()-2*TEXT_FIELD_TEXT_PADDING, (int) getHeight()-2*TEXT_FIELD_TEXT_PADDING, PApplet.JAVA2D);
+		}
+		textCanvas.beginDraw();
+		textCanvas.stroke(TEXT_FIELD_TEXT_COLOR.getRGB());
+		textCanvas.fill(TEXT_FIELD_TEXT_COLOR.getRGB());
+		textCanvas.textAlign(PApplet.LEFT, PApplet.TOP);
+		textCanvas.textFont(NUIController.getInstance().getFont());
+		textCanvas.rectMode(PApplet.CORNER);
+		
+		textCanvas.background(TEXT_FIELD_FILL_COLOR.getRGB());
+		float caretPos = textCanvas.textWidth(getText().substring(0, getCaretIndex()));
+		if (caretPos + textOffset > textCanvas.width) {
+			textOffset -= caretPos + textOffset - textCanvas.width + 2;
+		}
+		if (caretPos + textOffset < 0) {
+			textOffset += -(caretPos + textOffset) + 2;
+		}
+		textCanvas.text(getText(), textOffset, 0);
+		textCanvas.noStroke();
+		textCanvas.fill(16, 16, 255, 80);
+		textCanvas.rectMode(PApplet.CORNERS);
+		float secondaryPos = textCanvas.textWidth(getText().substring(0, getSecondaryCaretIndex()));
+		textCanvas.rect(secondaryPos + textOffset, 0, caretPos + textOffset, textCanvas.height);
+		if (!keyboards.isEmpty() && ++timer%30 < 15) {
+			textCanvas.stroke(TEXT_FIELD_TEXT_COLOR.getRGB());
+			textCanvas.line(caretPos + textOffset, 0, caretPos + textOffset, textCanvas.height);
+		}
+		textCanvas.endDraw();
+		
+		p.image(textCanvas, 0, 0);
+	}
+
+	public void keyPressed(KeyEvent e) {
+		if (e.isCoded()) {
+			switch (e.getKeyCode()) {
+				case KeyEvent.LEFT:
+					int i = getCaretIndex();
+					int s = getSecondaryCaretIndex();
+					if (shiftDown) {
+						setCaretIndex(i-1);
+					}
+					else {
+						if (i == s) {
+							setCaretIndex(i-1);
+							setSecondaryCaretIndex(getCaretIndex());
+						}
+						else if (i < s) {
+							setSecondaryCaretIndex(i);
+						}
+						else if (i > s) {
+							setCaretIndex(s);
+						}
+					}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 175:186, 194:205.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.